File: Python/SettingRTCConf/MainWindow.py
#!/bin/env python
# -*- encoding: utf-8 -*-

##
#   @file .py
#   @brief 



import thread


import optparse
import sys,os,platform
import re
import time
import random
import commands
import math
import imp



import RTC
import OpenRTM_aist

from OpenRTM_aist import CorbaNaming
from OpenRTM_aist import RTObject
from OpenRTM_aist import CorbaConsumer
from omniORB import CORBA
import CosNaming

from PyQt4 import QtCore, QtGui

from ConfigWidget import ConfigWidget
from CorbaWidget import CorbaWidget
from ExecCxtWidget import ExecCxtWidget
from LoggerWidget import LoggerWidget
from ManagerControl import ManagerControl
from ManagerWidget import ManagerWidget
from NamingWidget import NamingWidget
from TimerWidget import TimerWidget



class MainWindow(QtGui.QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()
        

        self.tab_widget = QtGui.QTabWidget(self)
        self.setCentralWidget(self.tab_widget)

        self.createAction()
	self.createMenus()

	#self.mgrc = ManagerControl("")
	self.mgrc = None
	

	#self.mgrc.CreateComp("MyFirstComponent",[".\\MyFirstComponent"])
        #self.mgrc.CreateComp("MyFirstComponent",[".\\MyFirstComponent"])
        
    ##
    #アクションの作成の関数
    ##
    def createAction(self):

	self.newAct = QtGui.QAction("&New...",self)
	self.newAct.setShortcuts(QtGui.QKeySequence.New)
        self.newAct.triggered.connect(self.newFile)
        


	self.openAct = QtGui.QAction("&Open...",self)
        self.openAct.setShortcuts(QtGui.QKeySequence.Open)
        self.openAct.triggered.connect(self.open)


        self.saveAct = QtGui.QAction("&Save",self)
        self.saveAct.setShortcuts(QtGui.QKeySequence.Save)
        self.saveAct.triggered.connect(self.save)

    ##
    #メニューの作成の関数
    ##
    def createMenus(self):

	self.fileMenu = self.menuBar().addMenu("&File")
	self.fileMenu.addAction(self.newAct)
        self.fileMenu.addAction(self.openAct)
        self.fileMenu.addAction(self.saveAct)


    def createTabs(self):
        self.Tabs = []
        self.ManagerTab = ManagerWidget(self.mgrc)
        self.tab_widget.addTab(self.ManagerTab, u"マネージャ")
        self.Tabs.append(self.ManagerTab)
        self.CorbaTab = CorbaWidget(self.mgrc)
	self.tab_widget.addTab(self.CorbaTab, u"CORBA")
	self.Tabs.append(self.CorbaTab)
	self.ConfigTab = ConfigWidget(self.mgrc)
	self.tab_widget.addTab(self.ConfigTab, u"一般的")
	self.Tabs.append(self.ConfigTab)
	self.NamingTab = NamingWidget(self.mgrc)
	self.tab_widget.addTab(self.NamingTab, u"ネームサービス")
	self.Tabs.append(self.NamingTab)
	self.LoggerTab = LoggerWidget(self.mgrc)
	self.tab_widget.addTab(self.LoggerTab, u"ロガー")
	self.Tabs.append(self.LoggerTab)
	self.TimerTab = TimerWidget(self.mgrc)
	self.tab_widget.addTab(self.TimerTab, u"タイマ")
	self.Tabs.append(self.TimerTab)
	self.ExecCxtTab = ExecCxtWidget(self.mgrc, self.ManagerTab)
	self.tab_widget.addTab(self.ExecCxtTab, u"実行コンテキスト")
	self.Tabs.append(self.ExecCxtTab)
	

    ##
    #ファイル読み込みスロット
    ##
    def open(self):
        if self.mgrc == None:
            fileName = QtGui.QFileDialog.getOpenFileName(self,u"開く","","Config File (*.conf);;All Files (*)")
            if fileName.isEmpty():
                return
            ba = str(fileName.toLocal8Bit())
            #ba = ba.replace("/","\\")
            self.mgrc = ManagerControl(ba)
            self.createTabs()

            

            
        else:
            self.mesBox(u"既にコンフィギュレーションファイルは開いています")

    ##
    #ファイル保存のスロット
    ##
    def save(self):

	fileName = QtGui.QFileDialog.getSaveFileName(self,u"保存", "","Config File (*.conf);;All Files (*)")
	if fileName.isEmpty():
            return False

	ba = str(fileName.toLocal8Bit())

	f = open(ba, "w")

        for t in self.Tabs:
            for k,j in t.WidList.items():
                s = k + ": "
                v = ""
                if j["Type"] == ManagerControl.TextBox:
                    v += str(j["Widget"].text().toLocal8Bit())
                elif j["Type"] == ManagerControl.TextCombox or j["Type"] == ManagerControl.Combox:
                    
                    
                    if k == "manager.modules.load_path" or k == "manager.modules.preload" or k == "manager.components.precreate" or k == "corba.endpoints":
                        for c in range(0, j["Widget"].count()):
                            v += str(j["Widget"].itemText(c).toLocal8Bit())
                            if c < j["Widget"].count()-1:
                                v += ","
                    else:
                        v += str(j["Widget"].currentText().toLocal8Bit())
                elif j["Type"] == ManagerControl.SpinBox or j["Type"] == ManagerControl.DoubleSpinBox:
                    v += str(j["Widget"].value())
                if v != "":
                    s += v + "\n"
                    f.write(s)



        fname = os.path.basename(ba)
        name, ext = os.path.splitext(fname)
        dname = os.path.dirname(os.path.relpath(ba))

        for c in self.mgrc.mgr.getComponents():
            s = c.getCategory() + "." + c.get_sdo_id() + ".config_file: " + c.get_sdo_id() + ".conf\n"
            f.write(s)
            if dname == "":
                path = ".\\"+c.get_sdo_id() + ".conf"
            else:
                path = dname+"\\"+c.get_sdo_id() + ".conf"
            f2 = open(path, "w")

            
            cstes = c.get_configuration().get_active_configuration_set()
            s = "configuration.active_config: " + cstes.id + "\n"
            f2.write(s)
            
            for l in c.get_configuration().get_configuration_sets():
                for d in l.configuration_data:
                    s = "conf." + l.id + "." + d.name + ": " + d.value.value() + "\n"
                    f2.write(s)

            oEC = c.get_owned_contexts()[0]
            rate = oEC.get_rate()
            s = "exec_cxt.periodic.rate: " + str(rate) + "\n"
            f2.write(s)

            #s = "exec_cxt.periodic.type: " + "" + "\n"
            f2.close()
                
            



	f.close()

	

	

    ##
    #初期化のスロット
    ##
    def newFile(self):
        if self.mgrc == None:
            self.mgrc = ManagerControl("")
            self.createTabs()

            
            """for c in self.mgrc.mgr.getComponents():
                print c.get_sdo_id()
                for l in c.get_configuration().get_configuration_sets():
                    print l.id
                    print l.description
                    #prop = OpenRTM_aist.SDOPackage.NVList
                    #OpenRTM_aist.toProperties(prop,l.configuration_data)
                    #print prop
                    for d in l.configuration_data:
                        print d.name
                        print d.value.value()"""
                
        else:
            self.mesBox(u"既にコンフィギュレーションファイルは開いています")


        

    def mesBox(self, mes):
        msgbox = QtGui.QMessageBox( self )
        msgbox.setText( mes )
        msgbox.setModal( True )
        ret = msgbox.exec_()

File: C++/MainWindow.cpp

#include <QLabel>
#include <QtWidgets>
#include <rtm/CorbaNaming.h>
#include <rtm/CorbaConsumer.h>

#include "MainWindow.h"


MainWindow::MainWindow()
{
	tc = QTextCodec::codecForLocale();

	tab_widget = new QTabWidget(this);
	setCentralWidget(tab_widget);

	createAction();
	createMenus();

	mgrc = NULL;
	//mgrc = new ManagerControl("");
	//createTabs();

}

void MainWindow::open()
{
	

	if(mgrc == NULL)
	{
		QString fileName = QFileDialog::getOpenFileName(this,
								tc->toUnicode("開く"), "",
								tr("Config File (*.conf);;All Files (*)"));
		if (fileName.isEmpty())
			return;

		std::string ba = fileName.toLocal8Bit();
		mgrc = new ManagerControl(ba);
		createTabs();
	}
	else
	{
		mesBox("既にコンフィギュレーションファイルは開いています");
	}
}

void MainWindow::save()
{
	QString fileName = QFileDialog::getSaveFileName(this,
							tc->toUnicode("保存"), "",
							tr("Config File (*.conf);;All Files (*)"));
	if (fileName.isEmpty())
        return;

	std::string ba = fileName.toLocal8Bit();
	std::ofstream ofs( ba );

	for(int i=0;i < Tabs.size();i++)
	{
		for(int j=0;j < Tabs[i]->WidList.size();j++)
		{
			std::string s = Tabs[i]->WidList[j]->m_name;
			s += ": ";
			std::string v = "";
			if(Tabs[i]->WidList[j]->m_type == TextBox)
			{
				QLineEdit *c = dynamic_cast<QLineEdit *>(Tabs[i]->WidList[j]->m_widget);
				if(c)
				{
					v += c->text().toLocal8Bit();
				}
			}
			else if(Tabs[i]->WidList[j]->m_type == TextCombox || Tabs[i]->WidList[j]->m_type == Combox)
			{
				QComboBox *c = dynamic_cast<QComboBox *>(Tabs[i]->WidList[j]->m_widget);
				if(c)
				{
					if(Tabs[i]->WidList[j]->m_name == "manager.modules.load_path" || Tabs[i]->WidList[j]->m_name == "manager.modules.preload" || Tabs[i]->WidList[j]->m_name == "manager.components.precreate" || Tabs[i]->WidList[j]->m_name == "corba.endpoints")
					{
					
						for(int k = 0;k < c->count();k++)
						{
							v += c->itemText(k).toLocal8Bit();
							if(k < c->count()-1)
							{
								v += ",";
							}
						}
					}
					else
					{
						v += c->currentText().toLocal8Bit();
					}
				}
				
			}
			else if(Tabs[i]->WidList[j]->m_type == SpinBox || Tabs[i]->WidList[j]->m_type == DoubleSpinBox)
			{
				QSpinBox *c = dynamic_cast<QSpinBox *>(Tabs[i]->WidList[j]->m_widget);
				if(c)
				{
					v += coil::otos<int>(c->value());
				}
				else
				{
					QDoubleSpinBox *cd = dynamic_cast<QDoubleSpinBox *>(Tabs[i]->WidList[j]->m_widget);
					if(cd)
					{
						v += coil::otos<double>(cd->value());
					}
				}
			}

			if(v != "")
			{
				s += v;
				ofs << s << std::endl;
			}
			
		}
	}

	PathList p = PathList(ba);

	std::string tpt = p.dname;
	tpt += "/";
	tpt += p.name;
	tpt += ".bat";
	std::ofstream pf( tpt );

	pf << "cd " << p.inv_rdir << std::endl;
	pf << "rtcd -f " << p.dname << "/" << p.name << "." << p.cxt << std::endl;
	pf.close();

	std::vector<RTC::RtcBase*> bl = mgrc->mgr->getComponents();
	for(int i=0;i < bl.size();i++)
	{
		

		std::string path;

		if(p.dname == "")
		{
			path = "./";
			path += bl[i]->get_sdo_id();
			path += ".conf";
		}
		else
		{
			path = p.dname;
			path += "/";
			path += bl[i]->get_sdo_id();
			path += ".conf";
		}

		std::ofstream ofs2( path );


		std::string s = bl[i]->getCategory();
		s += ".";
		s += bl[i]->get_sdo_id();
		s += ".config_file: ";
		s += path;


		ofs << s << std::endl;



		SDOPackage::ConfigurationSet* cstes = bl[i]->get_configuration()->get_active_configuration_set();
		s = "configuration.active_config: ";
		s += cstes->id;
		
		ofs2 << s << std::endl;

		SDOPackage::ConfigurationSetList* cl = bl[i]->get_configuration()->get_configuration_sets();
		
		//SDOPackage::Configuration_ptr cl = bl[i]->get_configuration();
		//cl->get_configuration_sets;
		//RTC::Properties properties;
		
		//std::vector<std::string> nl = properties.propertyNames();

		for(int j=0;j < cl->length();j++)
		{
			//std::cout << cl->length() << std::endl;
			//RTC::CorbaConsumer<SDOPackage::ConfigurationSet> cs;
			//SDOPackage::ConfigurationSetList_var clist;
			//cs.setObject(clist[j]);

			
			SDOPackage::ConfigurationSet l = cl->get_buffer()[j];
			//SDOPackage::NVList d = l.configuration_data;
			RTC::Properties properties;
			NVUtil::copyToProperties(properties, l.configuration_data);
			std::vector<std::string> kl = properties.propertyNames();
			for(int k=0;k < kl.size();k++)
			{
				
				s = "conf.";
				s += l.id;
				s += ".";
				s += kl[k];
				s += ": ";
				s += properties.getProperty(kl[k]);

				ofs2 << s << std::endl;
				//std::cout << properties.getProperty(kl[k]) << std::endl;
			}
			//properties[0].c_str();
			/*for(int k=0;k < d.length();k++)
			{
				

				NVUtil::copyToProperties(properties, confSet->configuration_data);

				//CORBA::Char* tmp;
				//CORBA::Any an_any = d.get_buffer()[k].value;
				//an_any <<= tmp;
				//if(an_any >>= tmp)
				//s += CORBA::Any::to_string(d.get_buffer()[k].value);
				

				ofs2 << s << std::endl;
			}*/
		}
		RTC::CorbaConsumer<RTC::ExecutionContext> ec0;
		RTC::ExecutionContextList_var ecs = bl[i]->get_owned_contexts();
		ec0.setObject(ecs[0]);
		double rate = ec0->get_rate();

		s = "exec_cxt.periodic.rate: ";
		s += coil::otos<double>(rate);
		s += "\n";

		ofs2 << s << std::endl;
		
		
		
	}

	ofs.close();
	
	


}

void MainWindow::newFile()
{
	if(mgrc == NULL)
	{
		mgrc = new ManagerControl("");
		createTabs();
	}
	else
	{
		mesBox("既にコンフィギュレーションファイルは開いています");
	}
}

void MainWindow::createAction()
{
	newAct = new QAction(tr("&New..."),this);
	newAct->setShortcuts(QKeySequence::New);
    connect(newAct, SIGNAL(triggered()), this, SLOT(newFile()));


	openAct = new QAction(tr("&Open..."),this);
    openAct->setShortcuts(QKeySequence::Open);
    connect(openAct, SIGNAL(triggered()), this, SLOT(open()));


    saveAct = new QAction(tr("&Save"),this);
    saveAct->setShortcuts(QKeySequence::Save);
    connect(saveAct, SIGNAL(triggered()), this, SLOT(save()));

	
}

void MainWindow::createMenus()
{
	fileMenu = menuBar()->addMenu(tr("&File"));
	fileMenu->addAction(newAct);
    fileMenu->addAction(openAct);
    fileMenu->addAction(saveAct);
}

void MainWindow::createTabs()
{
	ManagerWidget *ManagerTab = new ManagerWidget(mgrc, this);
	tab_widget->addTab(ManagerTab, tc->toUnicode("マネージャ"));
	Tabs.push_back(ManagerTab);
	CorbaWidget *CorbaTab = new CorbaWidget(mgrc, this);
	tab_widget->addTab(CorbaTab, tc->toUnicode("CORBA"));
	Tabs.push_back(CorbaTab);
	ConfigWidget *ConfigTab = new ConfigWidget(mgrc, this);
	tab_widget->addTab(ConfigTab, tc->toUnicode("一般的"));
	Tabs.push_back(ConfigTab);
	NamingWidget *NamingTab = new NamingWidget(mgrc, this);
	tab_widget->addTab(NamingTab, tc->toUnicode("ネームサービス"));
	Tabs.push_back(NamingTab);
	LoggerWidget *LoggerTab = new LoggerWidget(mgrc, this);
	tab_widget->addTab(LoggerTab, tc->toUnicode("ロガー"));
	Tabs.push_back(LoggerTab);
	TimerWidget *TimerTab = new TimerWidget(mgrc, this);
	tab_widget->addTab(TimerTab, tc->toUnicode("タイマ"));
	Tabs.push_back(TimerTab);
	ExecCxtWidget *ExecCxtTab = new ExecCxtWidget(mgrc, ManagerTab, this);
	tab_widget->addTab(ExecCxtTab, tc->toUnicode("実行コンテキスト"));
	Tabs.push_back(ExecCxtTab);
}

void MainWindow::mesBox(std::string mes)
{
	QMessageBox msgBox(this);
	msgBox.setText(tc->toUnicode(mes.c_str()));
	msgBox.exec();
}
